{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VZTHAfKlib52"
   },
   "source": [
    "# ***Notebook to interactively generate Kaplan-Meier plots***\n",
    "\n",
    "**Notebook made by:** *Eduardo Reyes-Alvarez (Ph.D. candidate)*\n",
    "\n",
    "**Affiliation:** *Dr. Lois Mulligan's lab, Queen's University.*\n",
    "\n",
    "**Contact:** *eduardo_reyes09@hotmail.com*\n",
    "\n",
    "**Date of latest version:** August 05, 2023."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "h4_4QcKNh3D3",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## **Instructions**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JrcC-iQLiajG"
   },
   "source": [
    "***Summary:***\n",
    "\n",
    "This notebook allows the user to easily and interactively create Kaplan-Meier probability curves (survival, progression, recurrence) by using the KaplanMeierFitter library for Python. \n",
    "\n",
    "***Note on the current version (V03):***\n",
    "\n",
    "* This is a third working version that was built based on the METABRIC dataset available on the cBioPortal website (https://www.cbioportal.org/study/summary?id=brca_metabric).\n",
    "* This version of the notebook allows to easily generate a KM plot using the whole dataset or dividing the dataset into multiple subgroups based on any column/variable contained in the clinical file or in the RNA Sequencing file from the same study.\n",
    "* This version has not yet been tested for other datasets from cBioPortals or elsewhere and no tests have been done with files other than the mentioned above.\n",
    "* Future releases will be made addressing this: Fixing some interaction issues. Generalizing pre-processing of files to allow the use of other studies from cBioPortals (first) and other databases (second). Making a version of this notebook for Google Colab to run online.\n",
    "\n",
    "***How to use this notebook:***\n",
    "\n",
    "1. **Getting the files:** Go to the link for cBioportals, get the whole METABRIC study (click on the arrow pointing down, on the right side of the study name). This will download a *.tar.gz* compressed file, where you can find and extract the files ***data_clinical_patient.txt*** and any of the ***data_mrna_illumina...*** files. You can extract them anywhere (desktop or downloads), but preferably in the same folder as where the notebook is (if not, don't worry, the notebook will ask you to upload file(s)).\n",
    "2. ***Rename*** the clinical file to ***clinical.txt*** and the mRNA one to ***RNA.txt***.\n",
    "3. ***Open the notebook*** with JupyterLab, select \"Run\" in the top-left menu, then \"Run all cells\" and wait (~1min). Due to widget layout and compatibility issues, it is best to use JupyterLab.\n",
    "4. ***Look under*** the code block on \"Start plotting here!\".\n",
    "5. ***Plot!***: start by selecting a time variable (the METABRIC study has only Overall Survival), the column with the event to observe, fill the 0 and 1 widgets with the event (0 = No event/Living, 1 = Event/Died). After that you can plot the whole dataset by clicking the green button at the bottom, or select \"Use variable(s)\" to explore different variables and make subgroups. Once subgroups are selected by filling the widgets with tags or ranges for each variable, you can click the green button to observe the multiple survival probabilities (please avoid overlap in the ranges, even if the numbers are integers, do this instead: 0.00-**30.00**, **30.01**-60.00).\n",
    "6.  ***Save!***: Once you have generated a valid KM plot, the button below the green button will be enabled, so you can save a jpg containing the plot, and an excel file with relevant information in case you wish to make the plot in other software.\n",
    "   \n",
    "**NOTES:** This notebook automatically generates a ***MyLog.txt*** file in the working directory to help troubleshoothing (it rewrites itself if you don't move it or rename it after each notebook interaction). Currently, the button \"Use variable(s)\" works well the first time, but if you plot once, then add/reduce the number of variables you may have issues so click the \"No\" button first and then again the \"Use variable(s)\" button (this issue will be addressed in the next version). Also, if you do not need the RNA Seq data, the notebook works with the clinical file alone too. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iZuptf8siU3v",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## **Code**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "roUHPIUNjaWd",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Install and import required libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "KB31u0HwmAjf"
   },
   "source": [
    "Some of the required libraries for this notebook may need to be installed before we can import them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "lHm25NFDgoFh"
   },
   "outputs": [],
   "source": [
    "# General libraries for data handling\n",
    "import os\n",
    "!pip install numpy\n",
    "!pip install pandas\n",
    "!pip install collections\n",
    "!pip install openpyxl\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import OrderedDict\n",
    "import openpyxl \n",
    "\n",
    "# To edit and handle the output of each code cell\n",
    "from IPython.display import display, clear_output, HTML\n",
    "\n",
    "# To make plots and figures\n",
    "!pip install matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# To do KM survival analysis\n",
    "!pip install lifelines\n",
    "from lifelines import KaplanMeierFitter\n",
    "\n",
    "# To generate interactive widgets\n",
    "!pip install ipywidgets\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import Output\n",
    "from ipywidgets import interact, interactive, fixed, interact_manual, Dropdown, HBox, VBox, Layout, Label\n",
    "\n",
    "# To log relevant information during the runtime\n",
    "import logging\n",
    "\n",
    "# Clear the output of the cell\n",
    "clear_output()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following steps are to set up the logging file, so we can register some relevant information in case we need to troubleshoot. After each session with the notebook, a file called ***MyLog.txt*** will be created in the same directory where the notebook is stored. It is recommended that the user renames this file if a problem arises because it gets rewritten every time the notebook is re-run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure the logging settings\n",
    "logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')\n",
    "\n",
    "# Create a logger\n",
    "logger = logging.getLogger()\n",
    "\n",
    "# Clear the existing log file or create a new empty one\n",
    "open('MyLog.txt', 'w').close()\n",
    "\n",
    "# Create a file handler\n",
    "file_handler = logging.FileHandler('MyLog.txt')\n",
    "file_handler.setLevel(logging.INFO)\n",
    "\n",
    "# Create a formatter and add it to the file handler\n",
    "formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')\n",
    "file_handler.setFormatter(formatter)\n",
    "\n",
    "# Add the file handler to the logger\n",
    "logger.addHandler(file_handler)\n",
    "\n",
    "# Log an initial message\n",
    "logger.info(f\"Log file created or cleared. \\n\")\n",
    "clear_output()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "inoo--ejg6g0",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Upload datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HN1F49MOj6VO"
   },
   "source": [
    "Request the user to upload:\n",
    "\n",
    "**1.** File with clinical data (named: ***clinical.txt***).\n",
    "\n",
    "**2.** -Optional- file with RNA Seq data (named: ***RNA.txt***).\n",
    "\n",
    "**NOTE:**  If the kernel is restarted and the file(s) already exist in the working directory (where the notebook is stored), the notebook will locate the files automatically and won't request their upload (to save time)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "x5fqV04Ch64p"
   },
   "outputs": [],
   "source": [
    "# Function that checks if the files required are in the directory or prompts the user to upload them\n",
    "def upload_input_files():\n",
    "\n",
    "    # Check if the clinical file is already in the working directory\n",
    "    if \"clinical.txt\" in os.listdir():\n",
    "        # Read the clinical file and log it\n",
    "        df_clinical = pd.read_csv(\"clinical.txt\", sep=\"\\t\", comment=\"#\")\n",
    "        logger.info(f\"Uploaded: clinical.txt \\n\")\n",
    "        clear_output()\n",
    "        \n",
    "        # Check if there is also an RNA file in the directory\n",
    "        if \"RNA.txt\" in os.listdir():\n",
    "            # Read the RNA file\n",
    "            df_RNA = pd.read_csv(\"RNA.txt\", sep=\"\\t\")\n",
    "            logger.info(f\"Uploaded: RNA.txt \\n\")\n",
    "            clear_output()\n",
    "        else:\n",
    "            df_RNA = None           \n",
    "    else:\n",
    "        # Ask the user to select the files to upload\n",
    "        print(\"Select the file(s) to upload (all at once):\")\n",
    "        uploaded_files = files.upload()\n",
    "\n",
    "        # Read the clinical file\n",
    "        df_clinical = pd.read_csv(\"clinical.txt\", sep=\"\\t\", comment=\"#\")\n",
    "        logger.info(f\"Uploaded: clinical.txt \\n\")\n",
    "\n",
    "        # Check if an RNA file was uploaded, and read it if so\n",
    "        if \"RNA.txt\" in uploaded_files:\n",
    "            df_RNA = pd.read_csv(\"RNA.txt\", sep=\"\\t\")\n",
    "            logger.info(f\"Uploaded: RNA.txt \\n\")\n",
    "        else:\n",
    "            df_RNA = None\n",
    "        \n",
    "    clear_output()\n",
    "\n",
    "    # Return the dataframes for the next steps\n",
    "    return df_clinical, df_RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Ebj9s47rseJW",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Pre-process uploaded files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "nGH-gL0e1-_3"
   },
   "source": [
    "This function currently processes a clinical file and a RNA file (optional) in the following way:\n",
    "\n",
    "* Some columns of the clinical dataset are searched and re-ordered based on relevant information that could be used for a KM plot, including: ***VITAL_STATUS, OS_STATUS + OS_MONTHS, PFS_STATUS + PFS_MONTHS, RFS_STATUS + RFS_MONTHS***. These columns appear in the METABRIC clinical dataset and likely have similar names in other studies (otherwise adjustments may be needed).\n",
    "\n",
    "* The METABRIC RNA dataset has all the genes measured as rows and patient IDs as columns, so the df is transposed and sorted in both axes so we can merge it later with the clinical df. Before doing that we also removed an unnecesary column with gene ID number (Entrez_Gene_Id)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "yaQag70xssyZ"
   },
   "outputs": [],
   "source": [
    "# This function searches OS/PFS/RFS _STATUS and _MONTHS columns in the clinical data\n",
    "# This function transposes the RNA dataset to have gene names as columns and patient IDs as rows (like the clinical data)\n",
    "def file_preprocessing(df_clinical, df_RNA):\n",
    "\n",
    "    ################### Processing for the clinical dataframe ##################\n",
    "\n",
    "    # Log the original dataframe\n",
    "    logger.info(f\"Preview of the original clinical dataset: \\n {df_clinical.iloc[:15, :10].to_string()} \\n\")\n",
    "    logger.info(f\"Data types of columns in the original clinical dataset: \\n {df_clinical.dtypes.to_string()} \\n\\n\")\n",
    "    clear_output()\n",
    "\n",
    "    # Prepare the variable with the reordered column names\n",
    "    clinical_columns_main = [\"PATIENT_ID\"]\n",
    "    time_to_event = []\n",
    "    event_observation = []\n",
    "\n",
    "    # Search for a column of Vital status (this is optional and may provide more information)\n",
    "    if \"VITAL_STATUS\" in df_clinical.columns:\n",
    "        clinical_columns_main.append(\"VITAL_STATUS\")\n",
    "        event_observation.append(\"VITAL_STATUS\")\n",
    "\n",
    "    # Search for Overall Survival columns\n",
    "    if \"OS_MONTHS\" in df_clinical.columns and \"OS_STATUS\" in df_clinical.columns:\n",
    "        clinical_columns_main.append(\"OS_MONTHS\")\n",
    "        time_to_event.append(\"OS_MONTHS\")\n",
    "        clinical_columns_main.append(\"OS_STATUS\")\n",
    "        event_observation.append(\"OS_STATUS\")\n",
    "\n",
    "    # Search for Recurrence-Free Survival columns\n",
    "    if \"RFS_MONTHS\" in df_clinical.columns and \"RFS_STATUS\" in df_clinical.columns:\n",
    "        clinical_columns_main.append(\"RFS_MONTHS\")\n",
    "        time_to_event.append(\"RFS_MONTHS\")\n",
    "        clinical_columns_main.append(\"RFS_STATUS\")\n",
    "        event_observation.append(\"RFS_STATUS\")\n",
    "\n",
    "    # Search for Progression-Free Survival columns\n",
    "    if \"PFS_MONTHS\" in df_clinical.columns and \"PFS_STATUS\" in df_clinical.columns:\n",
    "        clinical_columns_main.append(\"PFS_MONTHS\")\n",
    "        time_to_event.append(\"PFS_MONTHS\")\n",
    "        clinical_columns_main.append(\"PFS_STATUS\")\n",
    "        event_observation.append(\"PFS_STATUS\")\n",
    "\n",
    "    # Order alphabetically the remaining columns\n",
    "    clinical_columns_extra = [col for col in df_clinical.columns if col not in clinical_columns_main]\n",
    "    clinical_columns_extra.sort()\n",
    "\n",
    "    # Apply the re-ordering to the df\n",
    "    clinical_columns_ordered = clinical_columns_main + clinical_columns_extra\n",
    "    df_clinical = df_clinical[clinical_columns_ordered] \n",
    "\n",
    "    # Log the re-arranged dataframe\n",
    "    logger.info(f\"Preview of the pre-processed clinical dataset: \\n {df_clinical.iloc[:15, :10].to_string()} \\n\")\n",
    "    logger.info(f\"Data types of columns in the pre-processed clinical dataset: \\n {df_clinical.dtypes.to_string()} \\n\\n\")\n",
    "    clear_output()\n",
    "\n",
    "    ##################### Processing for the RNA dataframe #####################\n",
    "\n",
    "    # If an RNA file was uploaded, then the df is not empty\n",
    "    if df_RNA is not None:\n",
    "        # Log the original dataframe\n",
    "        logger.info(f\"Preview of the original RNA dataset: \\n {df_RNA.iloc[:15, :10].to_string()} \\n\")\n",
    "        logger.info(f\"Data types of some columns in the original RNA dataset: \\n {df_RNA.iloc[:, :10].dtypes.to_string()} \\n\\n\")\n",
    "        clear_output()\n",
    "        \n",
    "        # Drop the \"Entrez_Gene_Id\" column\n",
    "        df_RNA.drop(\"Entrez_Gene_Id\", axis=1, inplace=True)\n",
    "\n",
    "        # Rename the \"Hugo_Symbol\" column to \"PATIENT_ID\" as it appears in the clinical df\n",
    "        df_RNA.rename(columns={\"Hugo_Symbol\": \"PATIENT_ID\"}, inplace=True)\n",
    "\n",
    "        # Transpose the dataframe, making the content of the \"PATIENT_ID\" column the new column names\n",
    "        df_RNA = df_RNA.set_index(\"PATIENT_ID\").T\n",
    "\n",
    "        # Sort the gene names alphabetically\n",
    "        df_RNA.sort_index(axis=1, inplace=True)\n",
    "\n",
    "        # Reset the index to a numerical index\n",
    "        df_RNA = df_RNA.reset_index().rename_axis(\"\", axis=\"columns\")\n",
    "\n",
    "        # Rename the \"index\" column to \"PATIENT_ID\"\n",
    "        df_RNA.rename(columns={\"index\": \"PATIENT_ID\"}, inplace=True)\n",
    "\n",
    "        # Sort Patient IDs and reset the index\n",
    "        df_RNA = df_RNA.sort_values(\"PATIENT_ID\").reset_index(drop=True)\n",
    "\n",
    "        # Log the re-arranged dataframe\n",
    "        logger.info(f\"Preview of the pre-processed RNA dataset: \\n {df_RNA.iloc[:15, :10].to_string()} \\n\")\n",
    "        logger.info(f\"Data types of some columns in the pre-processed RNA dataset: \\n {df_RNA.iloc[:, :10].dtypes.to_string()} \\n\\n\")\n",
    "        clear_output()\n",
    "\n",
    "    ############################################################################\n",
    "\n",
    "    # Return the variables with the survival/progression/recurrence labels available in the datset\n",
    "    return df_clinical, df_RNA, time_to_event, event_observation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IpG4MvC6yArY",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Create interactive widgets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "kUWgV0v6yHxH"
   },
   "source": [
    "The general way to create interactive widgets is the following:\n",
    "\n",
    "**1.** Create the ***widget***.\n",
    "\n",
    "**2.** Create a ***widget output*** object (can be initially empty).\n",
    "\n",
    "**3.** Declare a ***handler function*** that is called when changes in the widget are made (here things can be shown in the widget output object).\n",
    "\n",
    "**4.** Call the ***observe*** attribute on the widget, specifying its handler function.\n",
    "\n",
    "**5.** ***Display*** the initial widget and initial state of the widget output.\n",
    "\n",
    "* Additionally, we can ***nest more subwidgets*** that appear when certain values/options in the main widget are selected, and ***each can have their own handler function and output object***. This is ideal for fine-tunning of complex interactions, however, the code can get very confusing nesting handler functions, observe/display/output calls inside others, in multiple levels. \n",
    "\n",
    "* An alternative approach used in this notebook is to first declare all the widget handler functions in the order they are called (with the exception of 1). Next, declare a general ***widget_preparation*** function that creates the widgets, their output objects, calls the observe attribute, and then displays all of them at once in a specific configuration. For this to work we had to set several variables as global and others to be passed with labmda functions when making the observe calls (since observe calls are not regular function calls with input/output parameters)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Time to Event handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once one of the options of the time to event dropdown is selected (***OS_MONTHS, PFS_MONTHS, RFS_MONTHS***), this function ***creates and displays a histogram*** so we can have a quick look at the information for all the patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of time_to_event_dropdown (histogram)\n",
    "def time_to_event_selection_handler(change):\n",
    "    \n",
    "    # Clear the output and return early if the default option is selected back \n",
    "    if change[\"new\"] == \"Click here to select...\":\n",
    "        with time_to_event_selection_info:\n",
    "            time_to_event_selection_info.clear_output()\n",
    "        return\n",
    "\n",
    "    # Check if the name selected is a column on the clinical dataframe\n",
    "    column_name = change[\"new\"]\n",
    "    if column_name in df_clinical.columns:\n",
    "        time_column = df_clinical[column_name].dropna()\n",
    "        logger.info(f\"The user selected: {column_name}     Widget: time_to_event_dropdown. \\n\")\n",
    "        logger.info(f\"Original dtype of {column_name}: {df_clinical[column_name].dtype}     Dtype once removing NANs: {time_column.dtype} \\n\")\n",
    "        \n",
    "        # Make histogram of values and handle exceptions\n",
    "        if time_column.dtype != \"object\":\n",
    "            plt.figure(figsize=(4, 2))\n",
    "            plt.hist(time_column, bins=12, color=\"darkgoldenrod\", ec=\"white\")\n",
    "            plt.xlabel(column_name)\n",
    "            plt.ylabel('Number of patients')\n",
    "            plt.title(f'Histogram of {column_name}')\n",
    "            logger.info(f\"A histogram was succesfully made and displayed for: {column_name} \\n\")\n",
    "        else:\n",
    "            info_str = \"Warning: Column type is not numeric.\"\n",
    "            logger.warning(\"User attention required: The time to event column may not be numerical. \\n\")\n",
    "    else:\n",
    "        info_str = \"Warning: Column not found in the dataframe.\"\n",
    "        logger.error(\"User attention required: The time to event column name was not found in the df. \\n\")\n",
    "\n",
    "    # Clear the output widget and display the additional information\n",
    "    with time_to_event_selection_info:\n",
    "        time_to_event_selection_info.clear_output()\n",
    "        plt.show() if time_column.dtype != \"object\" else print(info_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Event Observation handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once a selection has been made in the event observation dropdown (***VITAL_STATUS, OS_STATUS, PFS_STATUS, RFS_STATUS***), this function ***creates and displays a bar chart*** with the unique values found in the column selected. In addition, this function shows ***two other widgets** (tags input) that allow the user to assign the unique values that want to be ***labeled as 0 (No event was observed) and 1 (An event was observed)***. These selections are necessary and will be applied before passing the data to the KaplanMeierFitter objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of event_observation_dropdown (bar chart)\n",
    "def event_observation_selection_handler(change):\n",
    "\n",
    "    # Clear the output and return early if the default option is selected back \n",
    "    if change[\"new\"] == \"Click here to select...\":\n",
    "        event_observation_selection_info.clear_output()\n",
    "        event_observation_tagsinput_info.clear_output()\n",
    "        return\n",
    "        \n",
    "    # Check if the name selected is a column on the clinical dataframe\n",
    "    column_name = change[\"new\"]\n",
    "    if column_name in df_clinical.columns:\n",
    "        event_column = df_clinical[column_name]\n",
    "        logger.info(f\"The user selected: {column_name}     Widget: event_observation_dropdown. \\n\")\n",
    "        logger.info(f\"Dtype of {column_name}: {df_clinical[column_name].dtype}     Unique value counts: \\n\\t {event_column.value_counts(dropna=False).to_string()} \\n\")\n",
    "        \n",
    "        # Make a bar chart for unique values and handle exceptions\n",
    "        if event_column.dtype == \"object\":\n",
    "            value_counts = event_column.value_counts(dropna=False)\n",
    "            \n",
    "            if df_clinical[column_name].nunique() > 15:\n",
    "                logger.warning(\"User attention required: There may be something wrong with the event observation column as there are more than 15 unique values. \\n\")\n",
    "            \n",
    "            plt.figure(figsize=(4, 2))\n",
    "            value_counts.plot(kind='bar', color=['maroon', 'purple', 'green', 'crimson', 'navy', 'coral', 'lavender'])\n",
    "            plt.xlabel(column_name)\n",
    "            plt.xticks(rotation=45)\n",
    "            plt.ylabel('Number of patients')\n",
    "            plt.title(f'Bar Chart of {column_name}')\n",
    "            logger.info(f\"A bar chart was succesfully made and displayed for: {column_name} \\n\")\n",
    "        else:\n",
    "            info_str = \"Warning: Column type is not categorical.\"\n",
    "            logger.warning(\"User attention required: The event observation column may not be text-based. \\n\")\n",
    "    else:\n",
    "        info_str = \"Warning: Column not found in the dataframe.\"\n",
    "        logger.error(\"User attention required: The event observation column name was not found in the df. \\n\")\n",
    "\n",
    "    # Clear the output widget and display the additional information\n",
    "    with event_observation_selection_info:\n",
    "        event_observation_selection_info.clear_output()\n",
    "        plt.show() if event_column.dtype == \"object\" else print(info_str)\n",
    "    \n",
    "    # Following selection and plotting, get the unique values from the current column selection\n",
    "    event_values = np.ndarray.tolist(df_clinical[event_observation_dropdown.value].unique())\n",
    "\n",
    "    # Make subwidgets to specify the event to be observed so we can encode it in binary for kmf.fit\n",
    "    global event_observation_tagsinput0, event_observation_tagsinput1\n",
    "    event_observation_tagsinput0 = widgets.TagsInput(allowed_tags=event_values)\n",
    "    event_observation_tagsinput1 = widgets.TagsInput(allowed_tags=event_values)\n",
    "    event_observation_tagsinput = [widgets.HTML(\"<br/>\"), HBox([widgets.Label(\"0:\"), event_observation_tagsinput0]),\n",
    "                                   widgets.HTML(\"<br/>\"), HBox([widgets.Label(\"1:\"), event_observation_tagsinput1])]\n",
    "    event_observation_tagsinput_box = VBox(event_observation_tagsinput)\n",
    "    \n",
    "    with event_observation_tagsinput_info:\n",
    "        event_observation_tagsinput_info.clear_output()\n",
    "        display(event_observation_tagsinput_box)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Subgroup Options handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the button default value (***No***) is changed to ***Use variable(s)***, this function ***displays a slider*** that will make widget areas for each of the variables wanted. When the button default value is selected again, the ***outputs*** from the other button are all ***cleared up***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of subgroup_buttons (slider and widget areas based on the slider)\n",
    "def subgroup_selection_handler(change):\n",
    "\n",
    "    # The slider needs to be dynamically set/reset\n",
    "    global variable_number_slider, variable_repeats\n",
    "    \n",
    "    # If the user wants to make subgroups request a dataset, variable and number of groups (per variable)\n",
    "    if change['new'] == 'Use variable(s)':\n",
    "        logger.info(f\"The user selected: Use variable(s)     Widget: subgroup_buttons \\n\")\n",
    "        \n",
    "        # Show and observe changes in the multiple variable slider and add/remove widget areas as needed\n",
    "        variable_number_slider = widgets.IntSlider(min=1, max=5, value=1)\n",
    "        variable_number_slider.observe(variable_number_selection_handler, 'value')\n",
    "        display(HBox([widgets.HTML('\\u2003' * 4), widgets.Label(\"Make subgroups?:\"), subgroup_buttons, widgets.HTML('\\u2003' * 4), widgets.Label(\"Number of variables:\"), variable_number_slider]), display_id='variable_number_slider')\n",
    "\n",
    "        # Set the initial/minimum number of variable repeats and call the handler function (because 1 is the default)\n",
    "        variable_repeats = 1\n",
    "        variable_number_selection_handler({\"new\": variable_repeats})\n",
    "        \n",
    "    else:\n",
    "        logger.info(f\"The user selected: No     Widget: subgroup_buttons \\n\")\n",
    "        \n",
    "        # Clear the widget area(s) if the other button was selected previously\n",
    "        for i in reversed(range(5)):\n",
    "            subgroup_output_areas[i]['subgroup_maker1_info'].clear_output()\n",
    "            subgroup_output_areas[i]['subgroup_maker2_info'].clear_output()\n",
    "            subgroup_output_areas[i]['subgroup_options_selection_info'].clear_output()\n",
    "\n",
    "        # We do not need show the multiple variable slider when No is selected\n",
    "        display(HBox([widgets.HTML('\\u2003' * 4), widgets.Label(\"Make subgroups?:\"), subgroup_buttons]), display_id='variable_number_slider')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Variable Number handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function is manually triggered with the default (***1***) value of variables to do subgroups. Since the maximum number of widget areas and output objects are already created (***5***), this function ***calls the observe attribute and displays as many widget areas missing*** from the current number. In addition, if we selected a lower number than the previous one (changed our mind from 3 variables to 2, for example), this function ***removes the extra widget areas*** from the display (they are not deleted, just not shown anymore)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of variable_number_slider (widget areas)\n",
    "def variable_number_selection_handler(change):\n",
    "    \n",
    "    logger.info(f\"Number of variables to make subgroups selected: {change['new']} \\n\")\n",
    "    global variable_repeats\n",
    "\n",
    "    # Add more widget areas if the user increased the number of variables desired\n",
    "    if change['new'] > variable_repeats:\n",
    "    \n",
    "        # The number of areas to add is the new number minus the previous (so the missing are added instead of creating all again)\n",
    "        for repeat in range(variable_repeats, change['new']):\n",
    "            \n",
    "            # Reset and initialize the first widget in the widget area for each new variable\n",
    "            subgroup_widget_areas[repeat]['dataset_dropdown'].value='Click here to select...'\n",
    "            subgroup_widget_areas[repeat]['dataset_dropdown'].observe(lambda value, repeat=repeat: dataset_selection_handler(value, repeat), 'value')\n",
    "\n",
    "            # Clear the outputs of the new widget area and show the initial widget\n",
    "            subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            with subgroup_output_areas[repeat]['subgroup_options_selection_info']:\n",
    "                subgroup_output_areas[repeat]['subgroup_options_selection_info'].clear_output()\n",
    "                display(widgets.HTML(\"<hr>\"))\n",
    "                display(subgroup_widget_areas[repeat]['dataset_dropdown'])\n",
    "   \n",
    "    # Remove widget areas if the user decreased the number of variables desired\n",
    "    elif change[\"new\"] < variable_repeats:\n",
    "        \n",
    "        # Clear out any previous outputs in the widget areas to remove\n",
    "        for repeat in range(variable_repeats - 1, change['new'] - 1, -1):\n",
    "            subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            subgroup_output_areas[repeat]['subgroup_options_selection_info'].clear_output()\n",
    "\n",
    "    # Show the minimum widget repeats number (1) when the button is selected for the first time\n",
    "    else:\n",
    "        # Reset and initialize the first widget in the widget area\n",
    "        subgroup_widget_areas[0]['dataset_dropdown'].value='Click here to select...'\n",
    "        subgroup_widget_areas[0]['dataset_dropdown'].observe(lambda value, repeat=0: dataset_selection_handler(value, repeat), 'value')\n",
    "\n",
    "        # Clear the outputs of the new widget area and show the initial widget\n",
    "        subgroup_output_areas[0]['subgroup_maker1_info'].clear_output()\n",
    "        subgroup_output_areas[0]['subgroup_maker2_info'].clear_output()\n",
    "        with subgroup_output_areas[0]['subgroup_options_selection_info']:\n",
    "            subgroup_output_areas[0]['subgroup_options_selection_info'].clear_output()\n",
    "            display(widgets.HTML(\"<hr>\"))\n",
    "            display(subgroup_widget_areas[0]['dataset_dropdown'])\n",
    "        \n",
    "    # Get the current number of widget areas to create\n",
    "    variable_repeats = change[\"new\"]        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Dataset handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once a widget area is created, a dropdown is shown to select the dataset (***clinical or RNA***) where the variable of interest is. Once a selection is made, this function ***shows two other widgets*** to the right of the original, one is a dropdown or combobox to search for the variable/column of interest, and the other is a slider with the number of subgroups desired.\n",
    "\n",
    "**NOTE:** The observe call for this function has a lambda function to also pass the number of repeat, which indicates where exactly the two subwidgets need to be shown (area 1-5). This is needed because we are using this function to do the same thing for all the 5 possible areas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of dataset_dropdown (two subwidgets) \n",
    "def dataset_selection_handler(change, repeat):\n",
    "\n",
    "    # Reset the subgroup_number_slider\n",
    "    subgroup_widget_areas[repeat]['subgroup_number_slider'].value = 1\n",
    "    subgroup_widget_areas[repeat]['variables_dropdown'].value = 'Click here to select...'\n",
    "    if df_RNA is not None:\n",
    "        subgroup_widget_areas[repeat]['variables_combobox'].value = ''\n",
    "        logger.info(f\"There are RNA and clinical dataframes available to make subgroups. \\n\")\n",
    "\n",
    "    # Clear the previous outputs\n",
    "    subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "    subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "    subgroup_output_areas[repeat]['subgroup_options_selection_info'].clear_output()\n",
    "    \n",
    "    # Display a dropdown and slider for clinical variables (excluding PATIENT_ID column)\n",
    "    if change[\"new\"] == \"clinical\":\n",
    "        logger.info(f\"The {change['new']} dataset was selected to make subgroups. \\n\")\n",
    "        \n",
    "        with subgroup_output_areas[repeat]['subgroup_options_selection_info']:\n",
    "            subgroup_widget_areas[repeat]['variables_dropdown'].observe(lambda value, repeat=repeat: variables_selection_handler(value, repeat), 'value')\n",
    "            subgroup_widget_areas[repeat]['subgroup_number_slider'].observe(lambda value, repeat=repeat: group_number_selection_handler(value, repeat), 'value')\n",
    "            display(widgets.HTML(\"<hr>\"))\n",
    "            display(HBox([subgroup_widget_areas[repeat]['dataset_dropdown'], subgroup_widget_areas[repeat]['variables_dropdown'], widgets.HTML('\\u2003' * 3), subgroup_widget_areas[repeat]['subgroup_number_slider']]))\n",
    "    \n",
    "    # Display a combobox and slider for RNA variables (excluding PATIENT_ID column)\n",
    "    elif change[\"new\"] == \"RNA\":\n",
    "        logger.info(f\"The {change['new']} dataset was selected to make subgroups. \\n\")\n",
    "        \n",
    "        with subgroup_output_areas[repeat]['subgroup_options_selection_info']:\n",
    "            subgroup_widget_areas[repeat]['variables_combobox'].observe(lambda value, repeat=repeat: variables_selection_handler(value, repeat), 'value')\n",
    "            subgroup_widget_areas[repeat]['subgroup_number_slider'].observe(lambda value, repeat=repeat: group_number_selection_handler(value, repeat), 'value')\n",
    "            display(widgets.HTML(\"<hr>\"))\n",
    "            display(HBox([subgroup_widget_areas[repeat]['dataset_dropdown'], subgroup_widget_areas[repeat]['variables_combobox'], widgets.HTML('\\u2003' * 3), subgroup_widget_areas[repeat]['subgroup_number_slider']]))\n",
    "    \n",
    "    # Clear the output and display the original dropdown if the default option is selected back again\n",
    "    else:\n",
    "        logger.info(f\"The previous dataset to make subgroups was de-selected. \\n\")\n",
    "\n",
    "        with subgroup_output_areas[repeat]['subgroup_options_selection_info']:\n",
    "            display(widgets.HTML(\"<hr>\"))\n",
    "            display(subgroup_widget_areas[repeat]['dataset_dropdown'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Variable Selection handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once a dataset selection triggers the appearance of two other widgets, this function is called when a selection is made in the first of those. This function ***creates and displays a bar chart*** when the variable selected (column name) contains ***text*** data, or a ***histogram*** when the variable contains ***numbers***. Depending on the dataset selected, this function handles two possible widgets, either a dropdown (clinical dataset) or a combobox (RNA dataset).\n",
    "\n",
    "**NOTE:** The observe call for this function has a lambda function to also pass the number of repeat, which indicates where exactly the plots need to be shown (area 1-5). This is needed because we are using this function to do the same thing for all the 5 possible areas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of variables_dropdown and variables_combobox (plots)\n",
    "# Reminder that this function has to work for both df_clinical and df_RNA\n",
    "def variables_selection_handler(change, repeat):\n",
    "    \n",
    "    # Reset the subgroup_number_slider to remove any previous widget boxes\n",
    "    subgroup_widget_areas[repeat]['subgroup_number_slider'].value = 1\n",
    "    \n",
    "    # Display empty space if the default value on the dropdown or combobox is selected\n",
    "    if change[\"new\"] == 'Click here to select...' or change[\"new\"] == '':\n",
    "        logger.info(f\"The previous variable to make subgroups was de-selected. \\n\")\n",
    "        subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "        subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "        return\n",
    "    # If the user has selected a variable, do 4 steps: Apply event labels, find the column, plot, capture all columns of interest\n",
    "    else:\n",
    "        ##### Preparation part\n",
    "        # Make these variables global so they can be accessed from other subwidgets\n",
    "        global column_data, KM_data_1var, KM_data_all\n",
    "        KM_data_1var = df_clinical.copy()\n",
    "        \n",
    "        # This is to avoid plotting while incomplete strings are being searched in the combobox\n",
    "        if subgroup_widget_areas[repeat]['variables_combobox'] is not None and subgroup_widget_areas[repeat]['dataset_dropdown'].value == \"RNA\": \n",
    "            if change[\"new\"] not in subgroup_widget_areas[repeat]['variables_combobox'].options:\n",
    "                subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "                subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "                return\n",
    "\n",
    "        ##### Step 01 \n",
    "        # Apply the 0 and 1 labels to the event observed column and filter those values \n",
    "        if event_observation_tagsinput0.value and event_observation_tagsinput1.value:\n",
    "            for tag in event_observation_tagsinput0.value:\n",
    "                KM_data_1var[event_observation_dropdown.value] = KM_data_1var[event_observation_dropdown.value].replace(tag, \"0\")\n",
    "            for tag in event_observation_tagsinput1.value:\n",
    "                KM_data_1var[event_observation_dropdown.value] = KM_data_1var[event_observation_dropdown.value].replace(tag, \"1\")\n",
    "\n",
    "            # Filter the selected 0/1 event labels, transform column to integers and keep only the common first 3 columns \n",
    "            KM_data_1var = KM_data_1var.loc[KM_data_1var[event_observation_dropdown.value].isin([\"0\", \"1\"])]\n",
    "            KM_data_1var[event_observation_dropdown.value] = KM_data_1var[event_observation_dropdown.value].astype(int)\n",
    "\n",
    "        # Log the current status of KM_data_1var\n",
    "        logger.info(f\"[Subgrouping 1st step] The user selected to label -{str(event_observation_tagsinput0.value)}- as 0, and -{str(event_observation_tagsinput1.value)}- as 1. \\n\")\n",
    "        logger.info(f\"[Subgrouping 1st step] Apply 0/1 labels to column {event_observation_dropdown.value} on KM_data_1var: \\n {KM_data_1var.iloc[:15, :10].to_string()} \\n\")\n",
    "        logger.info(f\"[Subgrouping 1st step] Data types of KM_data_1var columns: \\n {KM_data_1var.dtypes.to_string()} \\n\\n\")\n",
    "\n",
    "        ##### Step 02\n",
    "        # Look for the selected column in either df, as it is not specified within this function\n",
    "        if change[\"new\"] in df_clinical.columns:\n",
    "            # Keep only the working columns, log it and extract the column to plot the values\n",
    "            KM_data_1var = KM_data_1var[['PATIENT_ID', time_to_event_dropdown.value, event_observation_dropdown.value, change[\"new\"]]]          \n",
    "            logger.info(f\"[Subgrouping 2nd step] The column {change['new']} -{KM_data_1var.dtypes[change['new']]} dtype- from df_clinical was selected to make subgroups. \\n\")\n",
    "            column_data[repeat] = KM_data_1var[change[\"new\"]].copy()\n",
    "            \n",
    "        # If the column is in df_RNA, joining is needed to combine it with the clinical columns\n",
    "        elif df_RNA is not None and change[\"new\"] in df_RNA.columns:\n",
    "            # Keep only the working columns from both dfs, log it and extract the column to plot the values\n",
    "            KM_data_1var = KM_data_1var[['PATIENT_ID', time_to_event_dropdown.value, event_observation_dropdown.value]]\n",
    "            df_RNA2 = df_RNA[['PATIENT_ID', change[\"new\"]]]\n",
    "            KM_data_1var = KM_data_1var.merge(df_RNA2, on='PATIENT_ID', how='inner')\n",
    "            logger.info(f\"[Subgrouping 2nd step] The column {change['new']} -{KM_data_1var.dtypes[change['new']]} dtype- from df_RNA was selected to make subgroups. \\n\")\n",
    "            column_data[repeat] = KM_data_1var[change[\"new\"]].copy()\n",
    "\n",
    "        # Log the current status of KM_data_1var \n",
    "        logger.info(f\"[Subgrouping 2nd step] Keep relevant columns of KM_data_1var and only rows with 0/1 event labels: \\n {KM_data_1var.iloc[:15, :10].to_string()} \\n\")\n",
    "        logger.info(f\"[Subgrouping 2nd step] Data types of KM_data_1var columns: \\n {KM_data_1var.dtypes.to_string()} \\n\\n\")\n",
    "\n",
    "        ##### Step 03\n",
    "        # Make and display a bar chart for text columns showing the counts for unique values\n",
    "        if column_data[repeat].dtype == 'object':\n",
    "            value_counts = column_data[repeat].value_counts(dropna=False)\n",
    "            fig, ax = plt.subplots(figsize=(5, 3))\n",
    "            value_counts.plot(kind='bar', color=['indigo', 'khaki', 'lightblue', 'salmon', 'sienna', 'silver', 'aquamarine', 'coral', 'teal', 'olive'])\n",
    "            ax.set_xlabel(change[\"new\"])\n",
    "            ax.set_ylabel('Count')\n",
    "            ax.set_title(f'Unique Value Counts for {change[\"new\"]}')\n",
    "            plt.xticks(rotation=45)\n",
    "\n",
    "            # Clear the specific space to show the new plot\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            with subgroup_output_areas[repeat]['subgroup_maker1_info']:\n",
    "                subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "                plt.show()\n",
    "                display(HTML('<span style=\"color: red;\">This plot shows rows with the 0 and 1 events especified above!</span>'))\n",
    "        \n",
    "        # Make and display a histogram of frequencies for numerical columns\n",
    "        else:\n",
    "            fig, ax = plt.subplots(figsize=(5, 3))\n",
    "            ax.hist(column_data[repeat], bins='auto', color=\"darkblue\", ec=\"white\")\n",
    "            ax.set_xlabel(change[\"new\"])\n",
    "            ax.set_ylabel('Frequency')\n",
    "            ax.set_title(f'Histogram for {change[\"new\"]}')\n",
    "\n",
    "            # Clear the specific space to show the new plot\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            with subgroup_output_areas[repeat]['subgroup_maker1_info']:\n",
    "                subgroup_output_areas[repeat]['subgroup_maker1_info'].clear_output()\n",
    "                plt.show()\n",
    "                display(HTML('<span style=\"color: red;\">This plot shows rows with the 0 and 1 events especified above!</span>'))\n",
    "\n",
    "        ##### Step 04\n",
    "        # Pass the information of the current variable KM_data_1var so we can have all variables of interest in KM_data_all\n",
    "        if repeat == 0:\n",
    "            # The first variable gets just passed as it is after the steps above\n",
    "            KM_data_all = KM_data_1var.copy()\n",
    "        else:\n",
    "            # Check if the column is already in KM_data_all\n",
    "            if repeat + 3 in KM_data_all.columns:\n",
    "                # Replace the values in the existing column\n",
    "                KM_data_all.iloc[:, repeat + 3] = KM_data_1var[change[\"new\"]].copy()\n",
    "            else:\n",
    "                # Create a new column and assign the values\n",
    "                KM_data_all[change[\"new\"]] = KM_data_1var[change[\"new\"]].copy()\n",
    "        \n",
    "            # Rename the column to the desired name\n",
    "            KM_data_all.rename(columns={repeat + 3: change[\"new\"]}, inplace=True)\n",
    "\n",
    "        # Log the KM_data_all as we add/replace variables/columns of interest \n",
    "        logger.info(f\"[Subgrouping 2nd step] Updated KM_data_all with columns of interest: \\n {KM_data_all.iloc[:15, :10].to_string()} \\n\")\n",
    "        \n",
    "        # After the plot is made trigger the slider function to show two (the minimum) widgets in the box \n",
    "        subgroup_widget_areas[repeat]['subgroup_number_slider'].value = 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Group Number handler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once a dataset selection triggers the appearance of two other widgets, this function is called when a selection is made in the second of those (***int slider***). This function ***creates and displays a widget box*** containing repeats of either ***tags or float range*** widgets according to the number of groups selected in the slider. \n",
    "\n",
    "**NOTE:** The observe call for this function has a lambda function to also pass the number of repeat, which indicates where exactly the widget boxes need to be shown (area 1-5). This is needed because we are using this function to do the same thing for all the 5 possible areas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to display the output of group_number slider (widget boxes)\n",
    "def group_number_selection_handler(change, repeat):\n",
    "     \n",
    "    # This function uses the global variable column_data[repeat] created in the function below\n",
    "    global subgroup_boxes, subgroup_tagsinput, subgroup_floatrangeslider\n",
    "\n",
    "    # When the default value on the variable dropdown or combobox is selected, do not display subgrouping options\n",
    "    if subgroup_widget_areas[repeat]['variables_dropdown'].value == 'Click here to select...':\n",
    "        if df_RNA is None or subgroup_widget_areas[repeat]['variables_combobox'].value == '':\n",
    "            with subgroup_output_areas[repeat]['subgroup_maker2_info']:\n",
    "                subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "                display(HTML('<span style=\"color: red;\">Choose a variable first!</span>'))\n",
    "            return\n",
    "    \n",
    "    # Use tags to specify the desired groups for text columns (min 2, max 10 groups)\n",
    "    if change[\"new\"]>1 and column_data[repeat].dtype == 'object':\n",
    "        logger.info(f\"[Subgrouping 3rd step] The user selected to make {change['new']} subgroups with tags input labels. \\n\")\n",
    "\n",
    "        # Make as many tags input widgets and labels as selected and put them in their corresponding box\n",
    "        unique_values = np.ndarray.tolist(column_data[repeat].unique())\n",
    "        subgroup_tagsinput = [widgets.TagsInput(allowed_tags=unique_values, description=f'Group {i+1}') for i in range(change[\"new\"])]\n",
    "        subgroup_tagsinput_labels = [widgets.Label(value=f'Group {i+1}') for i in range(change[\"new\"])]\n",
    "        subgroup_boxes[repeat] = VBox([HBox([label, tagsinput]) for label, tagsinput in zip(subgroup_tagsinput_labels, subgroup_tagsinput)])\n",
    "\n",
    "        # Clear the specific space to show the widget box\n",
    "        with subgroup_output_areas[repeat]['subgroup_maker2_info']:\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            display(subgroup_boxes[repeat])\n",
    "            \n",
    "    # Use range sliders to specify the desired groups for numerical columns (min 2, max 10 groups)\n",
    "    elif change[\"new\"]>1:\n",
    "        logger.info(f\"[Subgrouping 3rd step] The user selected to make {change['new']} subgroups with float range sliders. \\n\")\n",
    "        \n",
    "        # Make as many float range slider widgets as selected and put them in their corresponding box\n",
    "        cleaned_column_data = column_data[repeat].dropna()\n",
    "        min_value = cleaned_column_data.min()\n",
    "        max_value = cleaned_column_data.max()\n",
    "        subgroup_floatrangeslider = [widgets.FloatRangeSlider(min=min_value, max=max_value, step=0.01, description=f'Group {i + 1}') for i in range(change[\"new\"])]\n",
    "        subgroup_boxes[repeat] = VBox(subgroup_floatrangeslider)\n",
    "\n",
    "        # Clear the specific space to show the widget box\n",
    "        with subgroup_output_areas[repeat]['subgroup_maker2_info']:\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            display(subgroup_boxes[repeat]) \n",
    "    \n",
    "    # Tell the user to remove the current variable if they want one group\n",
    "    else:\n",
    "        with subgroup_output_areas[repeat]['subgroup_maker2_info']:\n",
    "            subgroup_output_areas[repeat]['subgroup_maker2_info'].clear_output()\n",
    "            display(HTML('<span style=\"color: red;\">For 1 group just remove this variable!</span>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Displaying Widget Areas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once the file(s) are uploaded by the user and pre-processed by the other function above, this function is called with the dataset(s) provided and the time+event columns found in the clinical dataset. Then, it ***creates all the main widgets, their output objects, makes the observe calls to their respective handlers, and makes the layout to display everything (defined by the author)***.\n",
    "\n",
    "***NOTE:*** Most widgets are straightforward and are created with their output, then observe with their function (above), and are displayed. However, the gray buttons widget to make subgroups is the most complex one and required to make pre-defined widget areas and widget output areas (5), which are displayed always but will fill and update upon user interaction. There are three output areas: the first where 3 widgets display horizontally upon specific selections, then the other two are below of that area and split the screen in two, on the left side we have plots that provide information useful to decide on the ranges or labels of the subgroups to make, and on the right side we have the widgets that will help us select the specific values for each subgroup."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Main function to prepare and display the interactive widgets and subwidgets\n",
    "def widget_preparation(df_clinical, df_RNA, time_to_event, event_observation):\n",
    "\n",
    "    logger.info(f\"---------------User interaction with the widgets starts here--------------- \\n\")\n",
    "    clear_output()\n",
    "    \n",
    "    #################################### First widget - Time to event ########################################\n",
    "\n",
    "    global time_to_event_dropdown, time_to_event_selection_info\n",
    "    \n",
    "    # Create the widget \n",
    "    time_to_event_dropdown = widgets.Dropdown(options=[\"Click here to select...\"] + time_to_event)\n",
    "\n",
    "    # Create the widget output (plot)\n",
    "    time_to_event_selection_info = Output()\n",
    "\n",
    "    # Observe changes in the widget to call the function above            \n",
    "    time_to_event_dropdown.observe(time_to_event_selection_handler, names=\"value\")\n",
    "    \n",
    "    #################################### Second widget - Event observation ####################################\n",
    "\n",
    "    # Variables accessed by multiple functions\n",
    "    global event_observation_dropdown, event_observation_selection_info, event_observation_tagsinput_info\n",
    "    \n",
    "    # Create the widget \n",
    "    event_observation_dropdown = widgets.Dropdown(options=[\"Click here to select...\"] + event_observation)\n",
    "\n",
    "    # Create the widget outputs (plots and subwidgets)\n",
    "    event_observation_selection_info = Output()\n",
    "    event_observation_tagsinput_info = Output()\n",
    "    \n",
    "    # Observe changes in the widget to call the function above \n",
    "    event_observation_dropdown.observe(event_observation_selection_handler, names=\"value\")\n",
    "\n",
    "    #################### Third widget - Handle time and event columns that have other names ###################\n",
    "\n",
    "    # Create the main widget\n",
    "    event_observation_checkbox = widgets.Checkbox(description=\"Can't find your columns?\", value=False)\n",
    "\n",
    "    # This situation will be included in later releases, for now...\n",
    "    event_observation_checkbox = widgets.HTML(\"\")\n",
    "    \n",
    "    #################################### Fourth widget - Making subgroups #####################################\n",
    "\n",
    "    # Variables accessed by multiple functions\n",
    "    global subgroup_buttons, variable_number_slider, subgroup_widget_areas, subgroup_output_areas\n",
    "    global event_observation_tagsinput0, event_observation_tagsinput1, column_data, subgroup_boxes, KM_data_all\n",
    "    \n",
    "    # Create the main subgrouping widgets \n",
    "    subgroup_buttons = widgets.ToggleButtons(options=[\"No\", \"Use variable(s)\"])\n",
    "    variable_number_slider = widgets.HTML(\"\")\n",
    "\n",
    "    # Create 5 subwidget areas (more subwidgets) to use a maximum of 5 variables with this button (min 1)\n",
    "    subgroup_widget_areas = [{ 'dataset_dropdown' : widgets.Dropdown(options=['Click here to select...', 'clinical'] + (['RNA'] if df_RNA is not None else []),\n",
    "                                                                     value='Click here to select...', description='Dataset:'),\n",
    "                              'variables_dropdown' : widgets.Dropdown(options=['Click here to select...'] + list(df_clinical.columns[1:]), \n",
    "                                                                      value='Click here to select...', description='Variables:'),\n",
    "                              'variables_combobox' : widgets.Combobox(options=list(df_RNA.columns[1:]), placeholder='Type gene of interest here', \n",
    "                                                                      description='Genes:') if df_RNA is not None else None,\n",
    "                              'subgroup_number_slider' : widgets.IntSlider(min=1, max=10, description='Groups:', value=1) \n",
    "                              } for i in range(5)]      \n",
    "\n",
    "    # Create 5 subwidget output areas (one per area above)\n",
    "    subgroup_output_areas = [{'subgroup_options_selection_info' : Output(), \n",
    "                              'subgroup_maker1_info' : Output(),\n",
    "                              'subgroup_maker2_info' : Output()\n",
    "                             } for i in range(5)]\n",
    "\n",
    "    # Variables that need to be initialized with specific values/number of items\n",
    "    event_observation_tagsinput0 = None\n",
    "    event_observation_tagsinput1 = None\n",
    "    column_data = [\"0\", \"1\", \"2\", \"3\", \"4\"]\n",
    "    subgroup_boxes = [\"0\", \"1\", \"2\", \"3\", \"4\"]\n",
    "    KM_data_all = pd.DataFrame(columns=[\"0\", \"1\", \"2\"])\n",
    "    \n",
    "    # Observe changes in the main widget to call the rest if a change is detected  \n",
    "    subgroup_buttons.observe(subgroup_selection_handler, 'value') \n",
    "\n",
    "    #################################### Fifth widget - Calling KM Fitter #####################################\n",
    "    \n",
    "    # Variables accessed by multiple functions\n",
    "    global generate_plot_button, KM_plot_area\n",
    "    \n",
    "    # Create the widget \n",
    "    generate_plot_button = widgets.Button(description='Generate plot', disabled=False, button_style='success', \n",
    "                                          tooltip='Click me and wait for the plot to be displayed below!',\n",
    "                                          icon='chart-line') \n",
    "\n",
    "    # Create the widget output (plot)\n",
    "    KM_plot_area = Output()\n",
    "            \n",
    "    # Call the KM Fitter when the button is clicked\n",
    "    generate_plot_button.on_click(pass_KM_parameters)\n",
    "\n",
    "    ################################## Six widget - KM plot customization tools ###############################\n",
    "    \n",
    "    # Variables accessed by multiple functions\n",
    "    global CI_checkbox, plot_labels_checkbox\n",
    "    \n",
    "    # Create the main widget\n",
    "    CI_checkbox = widgets.Checkbox(description=\"Show Confidence Intervals\", value=True)\n",
    "    plot_labels_checkbox = widgets.Checkbox(description=\"Move legend to right side\", value=False)\n",
    "\n",
    "    #################################### Seventh widget - Saving results ######################################\n",
    "    \n",
    "    # Variables accessed by multiple functions\n",
    "    global save_button\n",
    "    \n",
    "    # Once any of the plots is made and displayed, we update the saving button to show it\n",
    "    save_button = widgets.Button(description='Save results', disabled=True, button_style='', \n",
    "                                 tooltip='Click and look for a jpg and an excel file in the current directory!',\n",
    "                                 icon='download') \n",
    "            \n",
    "    # Call the saving function when there is a plot and the save button appears\n",
    "    save_button.on_click(save_KM_results)\n",
    "                \n",
    "    ########################################## Display initial widgets ########################################\n",
    "\n",
    "    # First, second and third widgets go together in the first row\n",
    "    display(widgets.HTML(\"<br/>\"))\n",
    "    display(HBox([widgets.Label(\"Time to Event:\"), time_to_event_dropdown, widgets.HTML('\\u2003' * 7), widgets.Label(\"Event Observation:\"), event_observation_dropdown, event_observation_checkbox]))\n",
    "    \n",
    "    # The output of the first two widgets is displayed in the second row\n",
    "    time_to_event_selection_info.layout.width = '41%'\n",
    "    event_observation_selection_info.layout.width = '41%'\n",
    "    event_observation_tagsinput_info.layout.width = '18%'\n",
    "    display(HBox([time_to_event_selection_info, event_observation_selection_info, event_observation_tagsinput_info]))\n",
    "    display(widgets.HTML(\"<br/>\"))\n",
    "\n",
    "    # The fourth widget is displayed in the third row\n",
    "    display(HBox([widgets.HTML('\\u2003' * 4), widgets.Label(\"Make subgroups?:\"), subgroup_buttons]), display_id='variable_number_slider')\n",
    "    display(widgets.HTML(\"<br/>\"))\n",
    "\n",
    "    # Make room for the maximum widget areas (5)\n",
    "    for i in range(5):\n",
    "        display(subgroup_output_areas[i]['subgroup_options_selection_info'])\n",
    "        display(widgets.HTML(\"<br/>\"))\n",
    "        display(HBox([subgroup_output_areas[i]['subgroup_maker1_info'], subgroup_output_areas[i]['subgroup_maker2_info']]))\n",
    "    \n",
    "    # Finally, the button to start the KM Fitter and save the plot are diplayed in the sixth and seventg row\n",
    "    display(widgets.HTML(\"<br/>\"))\n",
    "    display(HBox([generate_plot_button, CI_checkbox, plot_labels_checkbox]))\n",
    "    display(save_button)\n",
    "    display(KM_plot_area)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Perform the KM analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function ***prepares and passes the data to the KM fitter*** whenever the green button is clicked on, and ***retrieves the KM estimates***. Ideally, the green button should be clicked when all the other wigets have some selections on them. This function handles the case when no subgroups are needed, and the cases when multiple subgroups are selected based on one or multiple variables (columns). \n",
    "\n",
    "In order to pass the data in the right format, when ***No*** subgroups are needed this function applies the labels selected for event observed and uses all the rows in the clinical dataset to calculate the estimates. On the other hand, when ***Using variable(s)*** is selected and one or more variables are chosen with 2 or more subgroups for each, this function applies the subgrouping labels on each column (the event labels were already applied), then filters out rows without the selected labels, and makes subsets of the daaset (which may combine clinical and RNA columns). The ***KM_analysis*** function is called and can work with any of these inputs, providing the estimates for each subset of the dataset, which are retrieved in this function and plotted.\n",
    "\n",
    "***NOTES:*** When plotting the KM estimates, this function will use 95% of the values to prevent the big drops or stretches that occurs when there are few samples left in the curve (the % can be easily adjusted)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to feed the current variable selections to the KM_analysis function (in the next subsection)\n",
    "def pass_KM_parameters(change):\n",
    "\n",
    "    # These variables will get the KMF objects and plots in all scenarios\n",
    "    global KM_analysis_output\n",
    "    \n",
    "    # If no subgrouping is required, apply the event observed tags and pass the data to KM_analysis\n",
    "    if subgroup_buttons.value == 'No':\n",
    "        \n",
    "        # Apply the selected labels on the event observation column \n",
    "        KM_data = df_clinical.copy()\n",
    "        if event_observation_tagsinput0.value and event_observation_tagsinput1.value:\n",
    "            for tag in event_observation_tagsinput0.value:\n",
    "                KM_data[event_observation_dropdown.value] = KM_data[event_observation_dropdown.value].replace(tag, \"0\")\n",
    "            for tag in event_observation_tagsinput1.value:\n",
    "                KM_data[event_observation_dropdown.value] = KM_data[event_observation_dropdown.value].replace(tag, \"1\")            \n",
    "        else:\n",
    "            with KM_plot_area:\n",
    "                display(HTML('<span style=\"color: red;\">Select first the values to label as 0 and 1 (No event, event)</span>'))\n",
    "            return\n",
    "\n",
    "        # Log the current status of KM_data\n",
    "        logger.info(f\"[No subgroups 1st step] The user selected to label -{str(event_observation_tagsinput0.value)}- as 0, and -{str(event_observation_tagsinput1.value)}- as 1. \\n\")\n",
    "        logger.info(f\"[No subgroups 1st step] Apply 0/1 labels to column {event_observation_dropdown.value} on KM_data: \\n {KM_data.iloc[:15, :10].to_string()} \\n\")\n",
    "        logger.info(f\"[No subgroups 1st step] Data types of KM_data columns: \\n {KM_data.dtypes.to_string()} \\n\\n\")\n",
    "                \n",
    "        # Filter out non-desired values and convert column to numbers for the KM Fitter\n",
    "        KM_data = KM_data[['PATIENT_ID', time_to_event_dropdown.value, event_observation_dropdown.value]]\n",
    "        KM_data = KM_data.loc[KM_data[event_observation_dropdown.value].isin([\"0\", \"1\"])]\n",
    "        KM_data[event_observation_dropdown.value] = KM_data[event_observation_dropdown.value].astype(int)\n",
    "\n",
    "        # Log the current status of KM_data\n",
    "        logger.info(f\"[No subgroups 2nd step] Keep relevant columns of KM_data and only rows with 0/1 event labels: \\n {KM_data.head(15).to_string()} \\n\")\n",
    "        logger.info(f\"[No subgroups 2nd step] Data types of KM_data columns: \\n {KM_data.dtypes.to_string()} \\n\\n\")\n",
    "        \n",
    "        # Pass the input parameters to the KM_analysis function and get back the KM objectt\n",
    "        KM_subgroups = []           \n",
    "        KM_analysis_output = KM_analysis(KM_data, KM_subgroups)\n",
    "\n",
    "        # Plot the estimate from the KMF object\n",
    "        with KM_plot_area:\n",
    "            KM_plot_area.clear_output()\n",
    "            plt.figure(figsize=(10, 6))\n",
    "            KM_analysis_output.plot(ci_show=CI_checkbox.value, legend=False, iloc=slice(0, int(len(KM_analysis_output.survival_function_) * 0.95)))\n",
    "            plt.xlabel(\"Time\")\n",
    "            plt.ylabel(\"Probability\")\n",
    "            plt.title(\"Kaplan-Meier Estimate\")\n",
    "            plt.show()\n",
    "            \n",
    "    ####################################################################\n",
    "    # If subgroups were selected, apply the corresponding tags or ranges\n",
    "    else:\n",
    "        # The dfs with all the information needed are these\n",
    "        global KM_data_all, subgroup_boxes\n",
    "\n",
    "        # Log the current status of KM_data_working\n",
    "        logger.info(f\"[Subgrouping 3rd step] Dataset KM_data_all before applying subgrouping labels: \\n {KM_data_all.head(15).to_string()} \\n\")\n",
    "        logger.info(f\"[Subgrouping 3rd step] Data types of KM_data_all before applying subgrouping labels: \\n {KM_data_all.dtypes.to_string()} \\n\\n\")\n",
    "        \n",
    "        # If the subgrouping changes are made multiple times, apply them to a copy of the original df\n",
    "        KM_data_working = KM_data_all.copy()\n",
    "\n",
    "        # Create an empty dictionary to store the mapping for each variable to reassign real group names\n",
    "        correct_group_labels = [{} for i in range(5)]\n",
    "\n",
    "        # Only iterate through the last number of variables in the slider (in case the user had more before)\n",
    "        for repeat in range(variable_repeats):\n",
    "\n",
    "            # To apply the tags in tagsinput we check if the VBoxes have repeats of label+tagsinput or floatrangesliders alone\n",
    "            if isinstance(subgroup_boxes[repeat].children[0], widgets.FloatRangeSlider):\n",
    "                \n",
    "                # Create a new column to store the group labels (original is numbers, new one will be text and next to the original)\n",
    "                KM_data_working.insert(repeat+4, 'TextSubgroup', '')\n",
    "\n",
    "                # Iterate through the float range sliders\n",
    "                for i, floatslider in enumerate(subgroup_boxes[repeat].children):\n",
    "                    # Retrieve the range selection and corresponding label\n",
    "                    subgroup_range = floatslider.value\n",
    "                    subgroup_label = floatslider.description\n",
    "                   \n",
    "                    # Get the indices of rows within the range selected\n",
    "                    subgroup_rows = (KM_data_working.iloc[:, repeat+3] >= subgroup_range[0]) & (KM_data_working.iloc[:, repeat+3] < subgroup_range[1])\n",
    "                \n",
    "                    # Assign the subgroup label to the matching rows\n",
    "                    KM_data_working.loc[subgroup_rows, 'TextSubgroup'] = subgroup_label\n",
    "\n",
    "                    # Add the correct label to the dictionary\n",
    "                    correct_group_labels[repeat][subgroup_label] = f\"{subgroup_range[0]:.2f} to {subgroup_range[1]:.2f}\"\n",
    "                \n",
    "                # Remove rows where the subgroup label is not assigned\n",
    "                KM_data_working = KM_data_working[KM_data_working['TextSubgroup'] != '']\n",
    "                variable_column_name = KM_data_working.columns[repeat+3]\n",
    "                KM_data_working.drop(variable_column_name, axis=1, inplace=True)\n",
    "                KM_data_working.rename(columns={'TextSubgroup': variable_column_name}, inplace=True)\n",
    "\n",
    "                # Log the ranges corresponding to each subgroup\n",
    "                log_string = \" - \".join([f\"Group {i+1}: {slider.value[0]:.2f} to {slider.value[1]:.2f}\" for i, slider in enumerate(subgroup_boxes[repeat].children)])\n",
    "                logger.info(f\"[Subgrouping 3rd step] Subgrouping labels applied to variable {repeat+1}---> {log_string} \\n\")\n",
    "\n",
    "            # The HBoxes containing tagsinput widgets and labels do not have the attribute lenght\n",
    "            else:\n",
    "                \n",
    "                # Retrieve the labels and elements to label\n",
    "                subgroup_selections = [tagsHBox.children[1].value for tagsHBox in subgroup_boxes[repeat].children]\n",
    "                subgroup_labels = [tagsHBox.children[0].value for tagsHBox in subgroup_boxes[repeat].children]\n",
    "                label_content_pairs = []\n",
    "                \n",
    "                # Iterate through the subgroup_selections list\n",
    "                for i, tags_list in enumerate(subgroup_selections):\n",
    "                    subgroup_elements = tags_list\n",
    "    \n",
    "                    # Generate a mapping of unique values to group labels\n",
    "                    element_to_label = {element: subgroup_labels[i] for element in subgroup_elements}\n",
    "\n",
    "                    # Add the group label to its list and the correct label to the dictionary\n",
    "                    label_content_pairs.extend([f\"{element}: {subgroup_labels[i]}\" for element in subgroup_elements])\n",
    "                    correct_group_labels[repeat][subgroup_labels[i]] = '+'.join(subgroup_elements)\n",
    "                    \n",
    "                    # Replace the subgroup elements with the new labels\n",
    "                    variable_column_name = KM_data_working.columns[repeat+3]\n",
    "                    KM_data_working[variable_column_name] = KM_data_working[variable_column_name].replace(element_to_label)\n",
    "            \n",
    "                # Filter out rows with new subgroup labels and log the labels selected\n",
    "                KM_data_working = KM_data_working[KM_data_working[variable_column_name].isin(subgroup_labels)]\n",
    "                log_string = \"  -  \".join(label_content_pairs)\n",
    "                logger.info(f\"[Subgrouping 3rd step] Subgrouping labels applied to variable {repeat+1}---> {log_string} \\n\")\n",
    "            \n",
    "            # Log the updated df\n",
    "            logger.info(f\"[Subgrouping 3rd step] Dataset KM_data_working after applying subgrouping labels: \\n {KM_data_working.head(15).to_string()} \\n\")\n",
    "        ########\n",
    "        # Once all labels have been applied to each column, make the subgroups\n",
    "\n",
    "        # Get the column indices for the extra columns\n",
    "        extra_column_indices = range(3, 3 + variable_repeats)\n",
    "        \n",
    "        # Get the unique values for each extra column\n",
    "        extra_column_unique_values = [KM_data_working.iloc[:, i].unique() for i in extra_column_indices]\n",
    "        \n",
    "        # Create an empty dictionary to store the subsets\n",
    "        KM_subgroups = {}\n",
    "        \n",
    "        # Generate all possible combinations of unique values from the extra columns\n",
    "        combinations = [[]]\n",
    "        for values in extra_column_unique_values:\n",
    "            combinations = [sublist + [value] for sublist in combinations for value in values]\n",
    "        \n",
    "        # Iterate through each combination of unique values\n",
    "        for combination in combinations:\n",
    "            # Create a subset of KM_data_working for the current combination\n",
    "            subset = KM_data_working.copy()\n",
    "            \n",
    "            # Filter the rows based on the current combination\n",
    "            for i, index in enumerate(extra_column_indices):\n",
    "                subset = subset[subset.iloc[:, index] == combination[i]]\n",
    "            \n",
    "            # Add the subset to the KM_subgroups dictionary\n",
    "            KM_subgroups[tuple(combination)] = subset\n",
    "\n",
    "        # Log the subgroups created\n",
    "        logger.info(f\"[Subgrouping 3rd step] Subgroups made from the dataset:\\n\")\n",
    "        for combination, subgroup in KM_subgroups.items():\n",
    "            logger.info(f\"Subgroup label: {combination}\")\n",
    "            logger.info(f\"\\n{subgroup.head(10)}\\n\")\n",
    "        \n",
    "        ########\n",
    "                \n",
    "        # Finally, pass the input parameters to the KM_analysis function and get back the KM object\n",
    "        KM_analysis_output = KM_analysis(KM_data_working, KM_subgroups)\n",
    "\n",
    "        # Reassign the real/correct labels (they are as Group X, and we will correct to the actual tag or range)\n",
    "        # Iterate through each key-value pair in the KM_analysis_output dictionary\n",
    "        for old_key, KM_object in list(KM_analysis_output.items()):\n",
    "            # Create a list to store the new key for this combination\n",
    "            new_key = []\n",
    "        \n",
    "            # Iterate through each element of the old key (a tuple of strings)\n",
    "            for i, label in enumerate(old_key):\n",
    "                # Retrieve the correct label from the corresponding dictionary in correct_group_labels\n",
    "                new_key.append(correct_group_labels[i].get(label, label))\n",
    "        \n",
    "            # Convert the new key (list) to a single string\n",
    "            new_key = ', '.join(new_key)\n",
    "        \n",
    "            # Replace the current key with the corrected_key in the KM_analysis_output dictionary\n",
    "            KM_analysis_output[new_key] = KM_analysis_output.pop(old_key)\n",
    "\n",
    "        \n",
    "        # Plot the estimates of all KMF objects (95% of data points)\n",
    "        with KM_plot_area:\n",
    "            KM_plot_area.clear_output()\n",
    "            plt.figure(figsize=(10, 6))\n",
    "            for label, KM_object in KM_analysis_output.items():\n",
    "                KM_object.plot(label=label, ci_show=CI_checkbox.value, iloc=slice(0, int(len(KM_object.survival_function_) * 0.95)))\n",
    "            plt.xlabel('Time')\n",
    "            plt.ylabel('Probability')\n",
    "            plt.title('Kaplan-Meier Estimates')\n",
    "            plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left') if plot_labels_checkbox.value else plt.legend()\n",
    "            plt.show()\n",
    "\n",
    "    # Once any of the plots is made and displayed, we enable the saving button to show it\n",
    "    save_button.description = 'Save results'\n",
    "    save_button.button_style = 'info'\n",
    "    save_button.disabled = False "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function ***feeds the input dataset(s) to the KaplanMeierFitter*** object, which can be a single instance or a dictionary containing an KMF object per subset of the dataset provided. It returns ***KMF_object*** which contains the estimates and can be plotted by the function above (pass_KM_parameters).\n",
    "\n",
    "***NOTES:*** The KM analysis currently does a simple .fit(time, event) of the data, but the resulting KMF objects have several properties that can be used to extract more insights or further complemented with other tools of the KMF library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def KM_analysis(KM_data, KM_subgroups):\n",
    "\n",
    "    # Unpack the input parameters provided\n",
    "    current_time_column = KM_data.columns[1]\n",
    "    current_event_column = KM_data.columns[2]\n",
    "    \n",
    "    global KMF_object\n",
    "\n",
    "    # Use the whole dataset when no groups were made\n",
    "    if subgroup_buttons.value == 'No':\n",
    "\n",
    "        # Create a single KaplanMeierFitter object\n",
    "        KMF_object = KaplanMeierFitter()\n",
    "\n",
    "        # Generate the plot using the specified columns\n",
    "        KMF_object.fit(durations=KM_data[current_time_column], event_observed=KM_data[current_event_column])\n",
    "\n",
    "        # Log part of the curve to verify the data was passed correctly\n",
    "        logger.info(f\"[No Subgroups 3rd step] The KM Fitter succesfully calculated the probabilities and made the plot. \\n\")\n",
    "        logger.info(f\"[No Subgroups 3rd step] Calculated survival function: \\n {KMF_object.survival_function_.head(7).to_string()} \\n ... \\n {KMF_object.survival_function_.tail(7).to_string()} \\n\\n\")\n",
    "\n",
    "    # Make a fit for every subset provided (based on the number of groups and subgroups made\n",
    "    else:\n",
    "\n",
    "        # Sort the subgroups in alphabetical order to plot them in the same order and colour\n",
    "        KM_subgroups = OrderedDict(sorted(KM_subgroups.items()))\n",
    "        \n",
    "        # Create an empty dictionary to store the KaplanMeierFitter objects\n",
    "        KMF_object = {}\n",
    "        logger.info(f\"[Subgrouping 4th step] The KM Fitter succesfully calculated the probabilities. \\n\")\n",
    "        \n",
    "        # Create KaplanMeierFitter objects for each subgroup in KM_subgroups\n",
    "        for label, subset in KM_subgroups.items():\n",
    "            kmf = KaplanMeierFitter()\n",
    "            kmf.fit(durations=subset[current_time_column], event_observed=subset[current_event_column])\n",
    "            KMF_object[label] = kmf\n",
    "\n",
    "            # Log part of the curve to verify the data was passed correctly\n",
    "            logger.info(f\"[Subgrouping 4th step] Calculated survival function of: {label}\")\n",
    "            logger.info(f\"\\n {kmf.survival_function_.head(7).to_string()} \\n ... \\n {kmf.survival_function_.tail(7).to_string()} \\n\\n\")\n",
    "        \n",
    "    return KMF_object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Save results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function first ***calls the helper function below to save the plot of interest to a .jpg image***, and then ***retrieves and saves the event table, survival function, confidence intervals, and median survival time of each KMF object***, putting all the information for each object in a formatted excel worksheet. The file is then saved as a .xlsx file in the current directory and then this "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "J2GaUdQjkhmK"
   },
   "outputs": [],
   "source": [
    "# Global counter to keep track of the number of times the button is clicked\n",
    "file_count = 1\n",
    "\n",
    "def save_KM_results(save_button):\n",
    "    global file_count\n",
    "\n",
    "    # Increment the file count to avoid overwriting previous plots and excel files\n",
    "    file_count_str = str(file_count).zfill(2)  # Convert the counter to a 2-digit zero-padded string\n",
    "    file_count += 1\n",
    "\n",
    "    # File names for Excel and plot\n",
    "    excel_filename = f\"KM_results_{file_count_str}.xlsx\"\n",
    "    plot_filename = f\"KM_results_{file_count_str}.jpg\"\n",
    "\n",
    "    # Save the plot image with the function below\n",
    "    save_plot_image(KM_analysis_output, plot_filename)\n",
    "\n",
    "    # Create a new Excel workbook and remove the default Sheet\n",
    "    workbook = openpyxl.Workbook()\n",
    "    workbook.remove(workbook['Sheet'])\n",
    "    \n",
    "    # Prepare the data to be processed (single KM object or list of KM objects)\n",
    "    if isinstance(KM_analysis_output, dict):\n",
    "        KM_objects_to_process = [{\"label\": f\"KM_Subgroup_{i+1}\", \"KM_object\": KM_object} for i, (label, KM_object) in enumerate(KM_analysis_output.items())]\n",
    "        real_labels = [f\"KM_Subgroup_{i+1}: {label}\" for i, (label, KM_object) in enumerate(KM_analysis_output.items())]\n",
    "    else:\n",
    "        # If KM_analysis_output is a single KM object, add it to the list as a dictionary with a general label\n",
    "        KM_objects_to_process = [{\"label\": \"KM_Dataset\", \"KM_object\": KM_analysis_output}]\n",
    "        real_labels = [\"KM_Dataset: Whole dataset - No subgroups\"]\n",
    "\n",
    "    # Process all KM curves/objects the same way \n",
    "    for index, data in enumerate(KM_objects_to_process):\n",
    "        \n",
    "        # Create a sheet per KM object\n",
    "        sheet = workbook.create_sheet(title=data[\"label\"])\n",
    "\n",
    "        # Write what the curve/object corresponds to\n",
    "        sheet.merge_cells(start_row=2, start_column=1, end_row=2, end_column=13)\n",
    "        sheet.cell(row=2, column=1).alignment = openpyxl.styles.Alignment(horizontal='center', vertical='center')\n",
    "        sheet.cell(row=2, column=1, value=real_labels[index]).font = openpyxl.styles.Font(bold=True, size=16)\n",
    "\n",
    "        # Get the tables from the KMF object\n",
    "        event_table = data[\"KM_object\"].event_table\n",
    "        survival_function = pd.DataFrame({\"Time\": data[\"KM_object\"].survival_function_.index,\n",
    "                                          \"Survival Probability\": np.ravel(data[\"KM_object\"].survival_function_.values)})\n",
    "        confidence_interval = data[\"KM_object\"].confidence_interval_\n",
    "        median_survival_time = data[\"KM_object\"].median_survival_time_\n",
    "\n",
    "        # Write the tables to the Excel sheet\n",
    "        tables = [event_table, survival_function, confidence_interval, median_survival_time]\n",
    "        table_names = [\"Event Table\", \"Survival Function\", \"Confidence Intervals\", \"Median Survival Time\"]\n",
    "        table_column_numbers = [1, 7, 10, 13]\n",
    "        \n",
    "        # Write all tables to the sheet\n",
    "        for col_index, (table, table_name) in enumerate(zip(tables, table_names)):\n",
    "            # Define the current column number for the table\n",
    "            current_column = table_column_numbers[col_index]\n",
    "    \n",
    "            # Set the header for the current table\n",
    "            sheet.cell(row=4, column=current_column, value=table_name).font = openpyxl.styles.Font(bold=True)\n",
    "    \n",
    "            if isinstance(table, pd.DataFrame):\n",
    "                # If the table is a DataFrame, convert it to a NumPy array\n",
    "                rows = table.to_numpy()\n",
    "                num_cols = len(table.columns)\n",
    "    \n",
    "                for row_index, row in enumerate(rows):\n",
    "                    # Write the data from the DataFrame to the Excel sheet\n",
    "                    for col_offset, value in enumerate(row):\n",
    "                        sheet.cell(row=row_index + 6, column=current_column + col_offset, value=value)  \n",
    "            else:\n",
    "                # If the table is not a DataFrame, write the single value to the Excel sheet\n",
    "                sheet.cell(row=5, column=current_column, value=table)\n",
    "\n",
    "        ##### Extra worksheet formatting \n",
    "        # Merge and center table titles\n",
    "        sheet.merge_cells(start_row=4, start_column=1, end_row=4, end_column=5)\n",
    "        sheet.cell(row=4, column=1).alignment = openpyxl.styles.Alignment(horizontal='center', vertical='center')\n",
    "        sheet.merge_cells(start_row=4, start_column=7, end_row=4, end_column=8)\n",
    "        sheet.cell(row=4, column=7).alignment = openpyxl.styles.Alignment(horizontal='center', vertical='center')\n",
    "        sheet.merge_cells(start_row=4, start_column=10, end_row=4, end_column=11)\n",
    "        sheet.cell(row=4, column=10).alignment = openpyxl.styles.Alignment(horizontal='center', vertical='center')\n",
    "        \n",
    "        # Write column titles and center them\n",
    "        sheet.cell(row=5, column=1, value=\"Removed\")\n",
    "        sheet.cell(row=5, column=2, value=\"Observed\")\n",
    "        sheet.cell(row=5, column=3, value=\"Censored\")\n",
    "        sheet.cell(row=5, column=4, value=\"Entrance\")\n",
    "        sheet.cell(row=5, column=5, value=\"At Risk\")\n",
    "        sheet.cell(row=5, column=7, value=\"Time\")\n",
    "        sheet.cell(row=5, column=8, value=\"Probability\")\n",
    "        sheet.cell(row=5, column=10, value=\"Lower Bound\")\n",
    "        sheet.cell(row=5, column=11, value=\"Upper Bound\")\n",
    "        for cell in [\"A5\", \"B5\", \"C5\", \"D5\", \"E5\", \"G5\", \"H5\", \"J5\", \"K5\"]:\n",
    "            sheet[cell].alignment = openpyxl.styles.Alignment(horizontal='center')\n",
    "        \n",
    "        # Adjust some column widths\n",
    "        for column, width in zip(['H', 'J', 'K', 'M'], [10, 12, 12, 22]):\n",
    "            sheet.column_dimensions[column].width = width\n",
    "        #####\n",
    "    \n",
    "    # Save the Excel file and log it\n",
    "    workbook.save(excel_filename)\n",
    "    logger.info(f\"An excel file containing the results has been saved to the current directory and the name {excel_filename} \\n\")\n",
    "    \n",
    "    # Update the button description\n",
    "    save_button.description = 'Results Saved!'\n",
    "    save_button.button_style = ''\n",
    "    save_button.disabled = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a helper function of the function above. This function ***makes again the KM plot to save it***, since the user may not want to save all the plots generated (like if exploratory analysis is done). For that reason, a button generates the plot, which can be either composed of a single curve (survival function) or by multiple, and another button does the saving of the data of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to save the plot image\n",
    "def save_plot_image(KM_analysis_output, filename):\n",
    "    \n",
    "    # Make the figure to fill\n",
    "    plt.figure(figsize=(10, 6))\n",
    "\n",
    "    if isinstance(KM_analysis_output, dict):\n",
    "        # Plot the estimates of all KMF objects (95% of data points)\n",
    "        for label, KM_object in KM_analysis_output.items():\n",
    "            KM_object.plot(label=label, ci_show=CI_checkbox.value, iloc=slice(0, int(len(KM_object.survival_function_) * 0.95)))\n",
    "    else:\n",
    "        # Plot the estimate from the single KMF object\n",
    "        KM_analysis_output.plot(ci_show=CI_checkbox.value, iloc=slice(0, int(len(KM_analysis_output.survival_function_) * 0.95)))\n",
    "    \n",
    "    # Customize the plot\n",
    "    plt.xlabel('Time')\n",
    "    plt.ylabel('Probability')\n",
    "    plt.title('Kaplan-Meier Estimates')\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left') if plot_labels_checkbox.value else plt.legend()\n",
    "\n",
    "    # Save, log and close the plot to release memory\n",
    "    plt.savefig(filename, dpi=600, bbox_inches='tight')\n",
    "    logger.info(f\"A jpg file containing the KM plot has been saved to the current directory and the name {filename} \\n\")\n",
    "    plt.close()  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "aeV8z3VakJmH",
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### Function for flow control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This function ***calls the functions to load, pre-process, and show the widgets***, so we can do all the steps and see all the outputs on the last code cell, which is when this function is called with a short line. This also ensures that other than the installation + import of libraries, we do not do anything until the very end, that way we avoid issues with variables not defined since not all the functions above are in strict order as they call each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "J9yore4pkLrj"
   },
   "outputs": [],
   "source": [
    "def start_plotting():\n",
    "\n",
    "    # Declare variables that need to be used by multiple functions\n",
    "    global df_clinical, df_RNA, time_to_event, event_observation\n",
    "    \n",
    "    # Load the input files\n",
    "    df_clinical, df_RNA = upload_input_files()\n",
    "\n",
    "    # Preprocess the uploaded files\n",
    "    df_clinical, df_RNA, time_to_event, event_observation = file_preprocessing(df_clinical, df_RNA)\n",
    "\n",
    "    # Generate and display the interactive widgets, these allow to select the data, make the KM plot(s) and save the results\n",
    "    widget_preparation(df_clinical, df_RNA, time_to_event, event_observation)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MEejaOVlitI7"
   },
   "source": [
    "## **Start KM-plotting here!!!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 53,
     "referenced_widgets": [
      "f85bbfe15c224a11bda855079c7db1d7",
      "0d117ab192664edb9aceea570dbfdb3e",
      "3bccde3637134f54adc7f14863743e3b",
      "48b4e258be964e5a8f9de3290b568ad6",
      "e4a1b7f6dfeb4c6eb541c8765a7c1602",
      "b138aa786e9d4cb4a605770f2142c19d",
      "6dbbc508d152463e8b20f231478d30cc",
      "1b651f79a4734168a797a7fce9a36a0a",
      "48923bcb11824c7a8a0def9eb0ea7b5a",
      "c1d87d0632234af9a79f0594c1458409",
      "152f0f3bd8a344bb9f1103c9a9aaf8ff",
      "98077bb27fd14b5ea1c69f2658692dcb",
      "f0d6949ca3ca454ab44478c2f987404b",
      "d8b47d0b629641d194e641e97fe1320b",
      "d5fc02eda8024cf58d8c1ee96bfe66b1",
      "9dd40f5abb8b4c9382f5f79b51b5cb93",
      "f5a4beaaf9f74ac18b2e434102bbe7d6",
      "75a828e351744905a28f357bccb2dc5b",
      "b702cc886f24435f9f27d22b903a094f",
      "1de0e09ce8744440ad4c5f0c58e0cb08",
      "381012fe4fb842d3ba10a615ca38c70b",
      "749f4a108d214c159e619e1ddbfec1f4",
      "6750cf84db574d94b4224ba6588237af",
      "247b8e942beb42baa530c95b16094f66",
      "47ac49cc34024e31ab8bc22bdbfd5b79",
      "3a78dfc073594d86bc7d4724889e1478"
     ]
    },
    "id": "PusHMSICi2bd",
    "outputId": "38ea75e8-b3f0-49b1-e330-87b97e7778a8"
   },
   "outputs": [],
   "source": [
    "# To begin, select: \"Run all cells\" and look under this cell\n",
    "start_plotting()"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [
    "h4_4QcKNh3D3",
    "roUHPIUNjaWd",
    "inoo--ejg6g0",
    "Ebj9s47rseJW",
    "IpG4MvC6yArY",
    "V7bNvpGykdAx",
    "aeV8z3VakJmH"
   ],
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "0d117ab192664edb9aceea570dbfdb3e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "VBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "VBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "VBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_48b4e258be964e5a8f9de3290b568ad6",
       "IPY_MODEL_e4a1b7f6dfeb4c6eb541c8765a7c1602"
      ],
      "layout": "IPY_MODEL_b138aa786e9d4cb4a605770f2142c19d"
     }
    },
    "152f0f3bd8a344bb9f1103c9a9aaf8ff": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "LabelModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "LabelModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "LabelView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_381012fe4fb842d3ba10a615ca38c70b",
      "placeholder": "​",
      "style": "IPY_MODEL_749f4a108d214c159e619e1ddbfec1f4",
      "value": "Event Observation:"
     }
    },
    "1b651f79a4734168a797a7fce9a36a0a": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DropdownModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DropdownModel",
      "_options_labels": [
       "Click here to select...",
       "OS_MONTHS"
      ],
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "DropdownView",
      "description": "Time to Event:",
      "description_tooltip": null,
      "disabled": false,
      "index": 1,
      "layout": "IPY_MODEL_f5a4beaaf9f74ac18b2e434102bbe7d6",
      "style": "IPY_MODEL_75a828e351744905a28f357bccb2dc5b"
     }
    },
    "1de0e09ce8744440ad4c5f0c58e0cb08": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "247b8e942beb42baa530c95b16094f66": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "381012fe4fb842d3ba10a615ca38c70b": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "3a78dfc073594d86bc7d4724889e1478": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "3bccde3637134f54adc7f14863743e3b": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "47ac49cc34024e31ab8bc22bdbfd5b79": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "48923bcb11824c7a8a0def9eb0ea7b5a": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "LabelModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "LabelModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "LabelView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_b702cc886f24435f9f27d22b903a094f",
      "placeholder": "​",
      "style": "IPY_MODEL_1de0e09ce8744440ad4c5f0c58e0cb08",
      "value": "Minimum value: 0.0\nMaximum value: 355.2"
     }
    },
    "48b4e258be964e5a8f9de3290b568ad6": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_6dbbc508d152463e8b20f231478d30cc",
       "IPY_MODEL_1b651f79a4734168a797a7fce9a36a0a",
       "IPY_MODEL_48923bcb11824c7a8a0def9eb0ea7b5a"
      ],
      "layout": "IPY_MODEL_c1d87d0632234af9a79f0594c1458409"
     }
    },
    "6750cf84db574d94b4224ba6588237af": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "6dbbc508d152463e8b20f231478d30cc": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "LabelModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "LabelModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "LabelView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_d5fc02eda8024cf58d8c1ee96bfe66b1",
      "placeholder": "​",
      "style": "IPY_MODEL_9dd40f5abb8b4c9382f5f79b51b5cb93",
      "value": "Time to Event:"
     }
    },
    "749f4a108d214c159e619e1ddbfec1f4": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "75a828e351744905a28f357bccb2dc5b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "98077bb27fd14b5ea1c69f2658692dcb": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DropdownModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DropdownModel",
      "_options_labels": [
       "Click here to select...",
       "OS_STATUS"
      ],
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "DropdownView",
      "description": "Event Observation:",
      "description_tooltip": null,
      "disabled": false,
      "index": 0,
      "layout": "IPY_MODEL_6750cf84db574d94b4224ba6588237af",
      "style": "IPY_MODEL_247b8e942beb42baa530c95b16094f66"
     }
    },
    "9dd40f5abb8b4c9382f5f79b51b5cb93": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "b138aa786e9d4cb4a605770f2142c19d": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "b702cc886f24435f9f27d22b903a094f": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "c1d87d0632234af9a79f0594c1458409": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "d5fc02eda8024cf58d8c1ee96bfe66b1": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "d8b47d0b629641d194e641e97fe1320b": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "e4a1b7f6dfeb4c6eb541c8765a7c1602": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_152f0f3bd8a344bb9f1103c9a9aaf8ff",
       "IPY_MODEL_98077bb27fd14b5ea1c69f2658692dcb",
       "IPY_MODEL_f0d6949ca3ca454ab44478c2f987404b"
      ],
      "layout": "IPY_MODEL_d8b47d0b629641d194e641e97fe1320b"
     }
    },
    "f0d6949ca3ca454ab44478c2f987404b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "LabelModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "LabelModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "LabelView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_47ac49cc34024e31ab8bc22bdbfd5b79",
      "placeholder": "​",
      "style": "IPY_MODEL_3a78dfc073594d86bc7d4724889e1478",
      "value": ""
     }
    },
    "f5a4beaaf9f74ac18b2e434102bbe7d6": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "f85bbfe15c224a11bda855079c7db1d7": {
     "model_module": "@jupyter-widgets/output",
     "model_module_version": "1.0.0",
     "model_name": "OutputModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/output",
      "_model_module_version": "1.0.0",
      "_model_name": "OutputModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/output",
      "_view_module_version": "1.0.0",
      "_view_name": "OutputView",
      "layout": "IPY_MODEL_3bccde3637134f54adc7f14863743e3b",
      "msg_id": "",
      "outputs": [
       {
        "name": "stdout",
        "output_type": "stream",
        "text": [
         "Minimum value: 0.0\n",
         "Maximum value: 355.2\n"
        ]
       }
      ]
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
