{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"collapsed_sections":["-7ysH-MIGHST","V3Cm8TcRGgcV","K5RP5mXgIefj","dxAQxew0I3fb","ClNv65NVbbyg","uxmU3YwaxQdt"]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"view-in-github"},"source":["<a href=\"https://colab.research.google.com/github/EdRey05/Resources_for_Mulligan_Lab/blob/main/Tools%20for%20students/Eduardo%20Reyes/04-ExtractCells_Broad_Institute_DepMapCCLE%5BColab%5D.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"]},{"cell_type":"markdown","metadata":{"id":"PAWt2popFuK9"},"source":["#***Notebook to extract data from the updated Cancer Cell Line Encyclopedia (DepMap Public 22Q2 files)***\n","\n","**Original data from:** *Broad Institute and Novartis*\n","\n","**Data downloaded from:** *https://depmap.org/portal/download/*\n","\n","**Notebook made by:** *Eduardo Reyes-Alvarez (Ph.D. candidate)*\n","\n","**Affiliation:** *Dr. Lois Mulligan's lab, Queen's University.*\n","\n","**Contact:** *eduardo_reyes09@hotmail.com*\n","\n","**Date of latest version:** December 08, 2022."]},{"cell_type":"markdown","metadata":{"id":"-7ysH-MIGHST"},"source":["##Instructions"]},{"cell_type":"markdown","metadata":{"id":"3LBKloWAGQyW"},"source":["***This is a similar version (with more recent data) to the notebook: 03-ExtractCells_Broad_Institute_CCLE_2019_[Colab]***\n","\n","***NOTE:*** You need to log in using the Gmail account for the Mulligan lab to use this notebook. Alternatively, you can download the files mentioned below from the website, save them in your Google Drive and edit the directories in the \"Main function\" section of the code.\n","\n","Before starting a search, it is recommended to check the website mentioned above and read the description of the files they have there. In this notebook, we used the most recent files (version 22Q2):\n","\n","  1.   **CCLE_expression.csv** : Gene expression TPM values of the protein coding genes for DepMap cell lines. Values are inferred from RNA-seq data using the RSEM tool and are reported after log2 transformation, using a pseudo-count of 1; log2(TPM+1).\n","  2.   **sample_info.csv** : Metadata for all of DepMapâ€™s cancer models/cell lines. A full description of each column is available in the DepMap Release README file.\n","\n","* You can download and try to preview the files in Excel if you want, but they are very big and may not load very well.\n","\n","* Once ready to run this notebook, place your cursor in the grey box under the \"Code\" section (it says 15 cells hidden), a \"play\" button should appear, click on it. After ~1min you can run the box under the \"Run your search\" section, a \"play\" button will appear, click on it and follow the instructions."]},{"cell_type":"markdown","metadata":{"id":"V3Cm8TcRGgcV"},"source":["##Code\n","\n"]},{"cell_type":"markdown","metadata":{"id":"K5RP5mXgIefj"},"source":["###Function to validate inputs"]},{"cell_type":"code","metadata":{"id":"4AWtoA5OHpuh"},"source":["#Input validation (type, min, max and range)\n","#Modified from here: https://stackoverflow.com/questions/23294658/asking-the-user-for-input-until-they-give-a-valid-response\n","\n","def check_input(prompt, type_=None, min_=None, max_=None, range_=None):\n","    if min_ is not None and max_ is not None and max_ < min_:\n","        raise ValueError(\"min_ must be less than or equal to max_.\")\n","    while True:\n","        user_input = input(prompt)\n","        if type_ is not None:\n","            try:\n","                user_input = type_(user_input)\n","            except ValueError:\n","                print(\"Input type must be {0}.\".format(type_.__name__))\n","                continue\n","        if max_ is not None and user_input > max_:\n","            print(\"Input must be less than or equal to {0}.\".format(max_))\n","        elif min_ is not None and user_input < min_:\n","            print(\"Input must be greater than or equal to {0}.\".format(min_))\n","        elif range_ is not None and user_input not in range_:\n","            if isinstance(range_, range):\n","                template = \"Input must be between {0.start} and {0.stop}.\"\n","                print(template.format(range_))\n","            else:\n","                template = \"Input must be {0}.\"\n","                if len(range_) == 1:\n","                    print(template.format(*range_))\n","                else:\n","                    expected = \" or \".join((\n","                        \", \".join(str(x) for x in range_[:-1]),\n","                        str(range_[-1])\n","                    ))\n","                    print(template.format(expected))\n","        else:\n","            return user_input"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"dxAQxew0I3fb"},"source":["###Function to search for cell lines"]},{"cell_type":"code","metadata":{"id":"aKDLd0DWtSyX"},"source":["def search_cells(cell_menu):\n","\n","  #Lists to store selected cell line names\n","  keepcells_name = []\n","\n","  #Loop to search text keys\n","  while True:\n","    #Check if we want to continue of exit\n","    continue_search = check_input(\"\\n Search for cell line? (Y/N) \", str, range_=(\"Y\", \"y\", \"N\", \"n\"))\n","    if continue_search==\"N\" or continue_search==\"n\":\n","      break\n","    \n","    #Get string of interest (we need a valid input to proceed)  \n","    while True:\n","      search_string = str(input(\"\\n Type term to search for: \"))\n","      search_string = search_string.upper()\n","      search_results = [cell for cell in cell_menu if search_string in cell]\n","      \n","      #Print columns that contain that string\n","      if search_results==[]:\n","        print(\"\\n Nothing was found! Try other term or a shorter version of it!\")\n","      else:\n","        search_results = [\"-\"] + search_results\n","        break\n","        \n","    for i,cell in enumerate(search_results):\n","          print(\" \\t \", i,cell)\n","    \n","    #Get index of cell line the user wants to extract and save its name\n","    keepcells_index = check_input(\"\\n Number of cell line to keep (use 0 if none are needed):\", int, 0, len(search_results)-1)\n","    if keepcells_index!=0:\n","      keepcells_name.append(search_results[keepcells_index]) \n","\n","    #Clear output window before starting at the top of the loop again\n","    clear_output(wait=True)\n","\n","  #Once the search is done, import all the columns selected by the user\n","  #We add back the column containing the gene names (Hugo_Symbol) and sort gene names\n","\n","  keepcells_name = sorted(keepcells_name)\n"," \n","  return keepcells_name"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ClNv65NVbbyg"},"source":["###Function to finish analysis and save"]},{"cell_type":"code","metadata":{"id":"7OjycBgmbfkG"},"source":["def end_analysis(extracted_RNA_data):\n","  \n","  if len(extracted_RNA_data.columns)>1:\n","    print(\"\\n \\t Preview of your dataset: \\n\", extracted_RNA_data.head())\n","    save_file = check_input(\"\\n Save dataset? (Y/N)\", str, range_=(\"Y\", \"y\", \"N\", \"n\"))\n","    \n","    if save_file==\"Y\" or save_file==\"y\":\n","      output_name = str(input(\"\\n Save file as: \"))\n","    \n","      global destination\n","      destination = data_output_dir + \"/\" + output_name + \".xlsx\"\n","      writer = pd.ExcelWriter(destination, engine='xlsxwriter')\n","      extracted_RNA_data.to_excel(writer, sheet_name=\"RNA_expression\")\n","      writer.save()\n","  \n","  new_search = check_input(\"\\n Start a new search? (Y/N)\", str, range_=(\"Y\", \"y\", \"N\", \"n\"))\n","  new_search = True if new_search==\"Y\" or new_search==\"y\" else False\n","  return new_search\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"uxmU3YwaxQdt"},"source":["###Main function"]},{"cell_type":"markdown","metadata":{"id":"DXQ7gpTxIq1H"},"source":["**Import required packages**"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QLhFNujRmWDT","outputId":"9b91c8be-b5aa-46c5-a2ca-9b8ffe5b1690"},"source":["#This is needed when working in Google Colab to synchronize google drive\n","from google.colab import drive\n","drive.mount('/content/drive/')\n","\n","import pandas as pd\n","from IPython.display import clear_output\n","\n","!pip install XlsxWriter"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount(\"/content/drive/\", force_remount=True).\n","Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Requirement already satisfied: XlsxWriter in /usr/local/lib/python3.8/dist-packages (3.0.3)\n"]}]},{"cell_type":"markdown","source":["**Import required files**"],"metadata":{"id":"j4ydYGO99pxT"}},{"cell_type":"code","metadata":{"id":"Hw02Q5W0PQjk"},"source":["#Directories of the 3 files to be used\n","directory1 = \"/content/drive/MyDrive/Colab Notebooks/Mulligan Lab/RNASeq/Broad_Institute_CCLE/input_files/DepMap Public 22Q2 Sample_info.csv\"\n","directory2 = \"/content/drive/MyDrive/Colab Notebooks/Mulligan Lab/RNASeq/Broad_Institute_CCLE/input_files/DepMap Public 22Q2 CCLE_expression.csv\"\n","\n","#Directory where any outputs generated will be saved\n","data_output_dir = \"/content/drive/MyDrive/Colab Notebooks/Mulligan Lab/RNASeq/Broad_Institute_CCLE/output_files\"\n","\n","#Import all csv files into dataframes\n","sample_IDs = pd.read_csv(directory1)\n","RNA_expression = pd.read_csv(directory2)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Pre-Processing**\n","\n","We need to do some pre-processing of the dataframes since the datasets are not exactly organized in the same way. The dataset with RNA expression has gene names as columns (which we want as rows) and the Achilles ID as rows (this code is the newer common identifier for all the datasets of DepMap) ***-the names of the cell lines are not given in this newer dataset-***. Since we want to generate a list of names of cell lines so the user can search their names, we need to open the Sample info file to get the CCLE Names (to be searched on), then get their corresponding Achilles ID so we can search them in the expression dataset and extract only the ones wanted by the user."],"metadata":{"id":"i4cN8sWq2Va9"}},{"cell_type":"code","source":["#First, sort the IDs by CCLE name and extract them to show the user the available cell lines (there are some NaN entries!!!)\n","sample_IDs = sample_IDs.sort_values(by=[\"CCLE_Name\"])\n","sample_IDs = sample_IDs[sample_IDs[\"CCLE_Name\"].notna()]\n","cell_menu = sample_IDs[\"CCLE_Name\"].values.tolist()\n","\n","#Second, the RNASeq dataset has no name for the Gene name column, and we need to transpose it\n","RNA_expression = RNA_expression.set_index(\"Unnamed: 0\").T\n","RNA_expression.index.name = \"Gene\"\n","RNA_expression = RNA_expression.sort_index(axis=1)\n","RNA_expression = RNA_expression.sort_index()\n"],"metadata":{"id":"MAzoVeA8xNa0"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["**Function to run the search loops**"],"metadata":{"id":"XyjQ-u4BEt6N"}},{"cell_type":"code","metadata":{"id":"Cco6uM-pU-9N"},"source":["def Begin_Search_Here():\n","\n","  #We do this loop in case the user wants to generate multiple files with different subsets of cell lines\n","  while True:\n","    \n","    #Pass the list of available cells, get back the ones requested by the user\n","    keepcells_name = search_cells(cell_menu)\n","    \n","    #Search the names the user wants in the sample info file to get their index and find their Achilles code\n","    #NOTE: The .index method gives Int64Index([1360], dtype='int64'), so we need to select the [0] item to get the number only\n","    keepcells_ACH = []\n","    for cell_name in keepcells_name:\n","      cell_name_index = sample_IDs.index[sample_IDs[\"CCLE_Name\"] == cell_name]\n","      keepcells_ACH.append(sample_IDs[\"DepMap_ID\"].loc[cell_name_index[0]])\n","    \n","    #Once we have the Achilles IDs, we can filter the dataset with the corresponding cell lines and rename them\n","    extracted_RNA_data = RNA_expression[keepcells_ACH]\n","    extracted_RNA_data.columns = keepcells_name\n","\n","    #Once done, we have processed this search, trigger the next one and/or save results\n","    clear_output(wait=True)\n","    new_search = end_analysis(extracted_RNA_data)\n","    if new_search==False:\n","      print(\"\\n Process completed! Your file(s) can be found in the output_files folder...\", \"\\n\\n To start a new search, run the Begin_Search_Here box again :) \")\n","      break\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UjZAXGFIGnG1"},"source":["##Run your search"]},{"cell_type":"code","metadata":{"id":"8fYY6u85nSA8"},"source":["#RUN THIS BOX TO START!\n","Begin_Search_Here()"],"execution_count":null,"outputs":[]}]}
